### PR TITLE
aborts backend requests using the abort-ch

### PR DESCRIPTION
**Pull request by shamsimam**
## Changes proposed in this PR

- aborts backend requests using the abort-ch

## Why are we making these changes?

Consistently abort backend requests using the `abort-ch` mechanism. Fallback to invoking `abort` directly only when the channel mechanism fails.

### CODE REVIEW

File: waiter/src/waiter/process_request.clj [hunk 1]
```diff
@@ -250,15 +250,29 @@
                                        (cid/with-correlation-id
                                          correlation-id
                                          (let [complete-trigger-id (utils/unique-identifier)]
+                                           (log/debug throwable "received from ctrl-chan" message)
                                            (deliver complete-triggered-promise complete-trigger-id)
                                            (when (= complete-trigger-id @complete-triggered-promise)
                                              (log/debug "closing request body as" message)
                                              (counters/dec! request-body-streaming-counter)
+                                             ;; the callback is necessary as there is a data race between aborting the
```
Comment (sradack) [inline waiter/src/waiter/process_request.clj]: Don't we control when `body-ch` gets closed later in the function?

File: waiter/src/waiter/process_request.clj [hunk 2]
```diff
@@ -250,15 +250,29 @@
...
+                                                                (cid/with-correlation-id
+                                                                  correlation-id
+                                                                  (log/info "request byte stream error, result of aborted request:" aborted?)
+                                                                  (async/close! body-ch)))]
+                                                 (log/error throwable "unable to stream request bytes, aborting request")
+                                                 (async/>!! abort-ch [throwable callback])))
                                              (report-request-size-metrics 0 true)
-                                             (async/close! body-ch))))))]
+                                             (if throwable
+                                               ;; TODO avoid the need for the timeout while sending to abort channel
```
Comment (sradack) [inline waiter/src/waiter/process_request.clj]: What would it take to avoid the timeout?

File: waiter/src/waiter/process_request.clj [hunk 3]
```diff
@@ -485,13 +499,19 @@
...
-                    (log/warn "unable to abort as request not found inside response!"))]
-      (log/info "aborted backend request:" aborted))))
+  [{:keys [abort-ch] :as response}]
+  (fn abort-backend-request-callback [^Exception ex]
+    (let [correlation-id (cid/get-correlation-id)
+          callback (fn abort-http-request-result-callback [aborted?]
+                     (cid/with-correlation-id
+                       correlation-id
+                       (log/debug ex "aborted backend request:" aborted? "due to" (.getClass ex))))]
+      (when abort-ch
```
Comment (sradack) [inline waiter/src/waiter/process_request.clj]: When do we not have an `abort-ch`?

File: waiter/src/waiter/process_request.clj [hunk 4]
```diff
@@ -485,13 +499,19 @@
...
+        (log/debug "requesting backend to be aborted via abort-ch"))
+      (when-not (and abort-ch (async/>!! abort-ch [ex callback]))
+        (log/debug "aborting backend request directly")
+        (let [aborted? (some-> response :request (.abort ex))]
```
Comment (sradack) [inline waiter/src/waiter/process_request.clj]: Should we log when we try to abort but are unable to?

File: waiter/src/waiter/process_request.clj [hunk 5]
```diff
@@ -250,15 +250,29 @@
...
+                                             ;; the abort request gets processed.
                                              (when throwable
-                                               (log/error throwable "unable to stream request bytes, aborting request")
-                                               (async/>!! abort-ch throwable))
+                                               (let [callback (fn complete-request-streaming-abort-result-callback [aborted?]
+                                                                (cid/with-correlation-id
+                                                                  correlation-id
+                                                                  (log/info "request byte stream error, result of aborted request:" aborted?)
+                                                                  (async/close! body-ch)))]
+                                                 (log/error throwable "unable to stream request bytes, aborting request")
```
Comment (sradack) [inline waiter/src/waiter/process_request.clj]:
```suggestion
                                                 (log/info throwable "unable to stream request bytes, aborting request")
```

File: waiter/src/waiter/process_request.clj (hunk 2 above)
Comment (shamsimam) [reply waiter/src/waiter/process_request.clj]: I created https://github.com/twosigma/waiter/pull/884 to show what we would need.

**Pull request closed without merging**